syntax = "proto2";
package oneflow.vm;

message MirroredParallelIdProto {}
message AllParallelIdProto {}

message OperandProto {
  required int64 logical_object_id = 1;
  oneof operand_type {
    int64 fixed_parallel_id = 2;
    MirroredParallelIdProto mirrored_parallel_id = 3;
    AllParallelIdProto all_parallel_id = 4;
  }
}

message OperandSeparatorProto { }

message InstructionOperandProto {
  oneof type {
    OperandProto const_operand = 1;
    OperandProto mutable_operand = 2;
    OperandProto mut2_operand = 3;
    OperandProto const_host_operand = 4;
    OperandProto mutable_host_operand = 5;
    OperandProto mut2_host_operand = 6;

    OperandSeparatorProto sep = 7;
    double double_i_operand = 8;
    int64 int64_i_operand = 9;
    uint64 uint64_i_operand = 10;
    bool bool_i_operand = 11;
  }
}

message InstructionProto {
  required string instr_type_name = 1;
  repeated InstructionOperandProto operand = 2;
};

message InstructionListProto {
  repeated InstructionProto instruction = 1;
}
